arangod/Cluster/ClusterComm.cpp

```diff
@@ -383,8 +386,22 @@ void ClusterComm::stopBackgroundThreads() {
   }  // for
 
   // pass 2:  verify each thread is stopped, wait if necessary
-  //          (happens in destructor)
+  //          No communication after this.
+  for (ClusterCommThread* thread : _backgroundThreads) {
+    if (!thread->runningInThisThread()) {
+      thread->halt();
+    }
+  }  // for
+}
+
+void ClusterComm::deleteBackgroundThreads() {
+  // pass 3:  de-alocate instances
+  // we want to keep the thread objects allocated till now,
+  // so eventual access to them doesn't fail.
   for (ClusterCommThread* thread : _backgroundThreads) {
+    if (thread->isRunning()) {
+      thread->halt();
```

Comment:
This `isRunning` check should be removed. If `thread::isRunning` returns false, that does not necessarily mean that the thread has already terminated and was joined or detached.  
And since `shutdown` is idempotent now, it is safe to call it multiple times.